hugr/src/algorithm/const_fold.rs

```
@@ -19,6 +22,90 @@ use crate::{
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Build a `ConstFoldConfig` with the given [VerifyLevel].
+    pub fn with_verify(mut self, verify: VerifyLevel) -> Self {
+        self.verify = verify;
+        self
+    }
+
+    fn verify_impl(
+        &self,
+        label: &str,
+        h: &impl HugrView,
+        reg: &ExtensionRegistry,
+    ) -> Result<(), ConstFoldError> {
+        match self.verify {
+            VerifyLevel::None => Ok(()),
+            VerifyLevel::WithoutExtensions => h.validate_no_extensions(reg),
+            VerifyLevel::WithExtensions => h.validate(reg),
+        }
+        .map_err(|err| ConstFoldError::verify_err(label, err))
+    }
+
+    /// Run the Constant Folding pass.
+    pub fn run(&self, h: &mut impl HugrMut, reg: &ExtensionRegistry) -> Result<(), ConstFoldError> {
+        self.verify_impl("input", h, reg)?;
+        loop {
+            // We can only safely apply a single replacement. Applying a
+            // replacement removes nodes and edges which may be referenced by
+            // further replacements returned by find_consts. Even worse, if we
+            // attempted to apply those replacements, expecting them to fail if
+            // the nodes and edges they reference had been deleted,  they may
+            // succeed because new nodes and edges reused the ids.
+            //
+            // We could be a lot smarter here, keeping track of `LoadConstant`
+            // nodes and only looking at their out neighbours.
+            let Some((replace, removes)) = find_consts(h, h.nodes(), reg).next() else {
+                break;
+            };
+            h.apply_rewrite(replace)?;
+            for rem in removes {
+                if let Ok(const_node) = h.apply_rewrite(rem) {
+                    // if the LoadConst was removed, try removing the Const too.
+                    let _ = h.apply_rewrite(RemoveConst(const_node));
```

Is it safe to ignore all errors here?